### DISCUSSION TITLE
Boto3 version Constraint needs to be updated. What to do?

### DISCUSSION THREAD

**Discussion by ghost** [apache/airflow#20340] [Q&A]
Hi, in the official constraint file for our setup (to be found here [constraints](https://raw.githubusercontent.com/apache/airflow/constraints-2.2.2/constraints-3.8.txt)) the version for boto3 is set to 1.18.65. 
This causes dependency conflicts because we also use aiobotocore which requires either newer versions or older versions than this specific boto3 release.

When will this constraint be updated and what can I do in the meantime?

**Comment by potiuk** [comment]
Constraints for 2.2.2 will not be updated. 

They are fixed during the 2.2.2 release and we very rarely update them (only when some unknown bug prevents installing airlfow). They are the "golden" set of dependencies that we tested airflow with when releasing it, but you are not limited to it. 

If you think it's safe you can upgrade botocore on your own (byt it's your own responsibility to test if it works). Constraints are just "constraints" they are not requirements. You are free to upgrade any dependencies you want (but then you have to make sure you know what you are doing and test if everything works as expected).

This is also described in https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html#constraints-files

> Airflow installation might be sometimes tricky because Airflow is a bit of both a library and application. Libraries usually keep their dependencies open and applications usually pin them, but we should do neither and both at the same time. We decided to keep our dependencies as open as possible (in setup.cfg and setup.py) so users can install different version of libraries if needed

BTW. When 2.2.3 will be released it will come with a new set of constraints: https://raw.githubusercontent.com/apache/airflow/constraints-2-2/constraints-3.8.txt  ~and there botocore is newer already (and 2.2.3 is tested with those dependencies). So the easiest way to get newer set of dependencies which are tested is to upgrade to 2.2.3 - if you follow [installation process](https://airflow.apache.org/docs/apache-airflow/stable/installation/index.html#using-pypi)  the constrained version of dependencies should be installed then~

**Reply by potiuk** [reply]
Correction. No Boto3 is the old version. Actually if you use amazon provider, currently the Amazon provider is limited. Not sure why but it is: 
https://github.com/apache/airflow/blob/6a77e849be3a505bf2636e8224862d77a4719621/setup.py#L190

```
amazon = [
    'boto3>=1.15.0,<1.19.0',
    'watchtower~=2.0.1',
    'jsonpath_ng>=1.5.3',
    'redshift_connector~=2.0.888',
    'sqlalchemy_redshift~=0.8.6',
    pandas_requirement,
]
```

This makes it possible that your upgrade will break because there is some breaking change. The way how to update is, is that someone who is interested (for example you) will make a PR and removes the limit and makes sure all the tests are passing (so fixes any breaking changes that occurs). Only then the new version will appear in constraints.

Airflow is maintained by many contributors - so the easiest way for you to make sure this is done, is to make a PR that removes the limit.

**Reply by potiuk** [reply]
BTW. It might well be that the limit was added by accident and your PR removing it might **just work** so I think you should try - at the worst case you will be able to ask community to help you with fixing the incompatibilities.

**Reply by ghost** [reply]
Alright thank you for your quick reply!
